src/gui/ar/areaCreator.js

```diff
@@ -0,0 +1,591 @@
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/.
+*/
+
+createNameSpace("realityEditor.gui.ar.areaCreator");
+
+/**
+ * @fileOverview realityEditor.gui.ar.areaCreator
+ * Provides an API for tools to call in order to prompt the user to draw an area on their screen which then gets
+ * returned to the tool
+ */
+
+realityEditor.gui.ar.areaCreator.Modes = {
+    DISABLED: "DISABLED",
+    DRAW_MODE_SELECT: "DRAW_MODE_SELECT",
+    AREA_CREATE: "AREA_CREATE",
+    HEIGHT_SET: "HEIGHT_SET",
+}
+realityEditor.gui.ar.areaCreator.mode = realityEditor.gui.ar.areaCreator.Modes.DISABLED;
+
+realityEditor.gui.ar.areaCreator.pointerCallbacks = {
+    down: () => {},
+    up: () => {},
+    move: () => {}
+}
+
+realityEditor.gui.ar.areaCreator.buttonCallbacks = {
+    cancel: () => {},
+    freehand: () => {},
+    polygon: () => {},
+    confirmArea: () => {},
+    confirmHeight: () => {},
+}
+
+realityEditor.gui.ar.areaCreator.polygonPoints = [];
+realityEditor.gui.ar.areaCreator.height = 1;
+
+realityEditor.gui.ar.areaCreator.canDragTouch = false; // Avoid initial click on menu for dragging
+realityEditor.gui.ar.areaCreator.lastFreehandTime = 0; // Prevents dropping too many points in freehand mode
+realityEditor.gui.ar.areaCreator.lastPointerY = null; // Prevent jumping between points by tapping twice
+
+realityEditor.gui.ar.areaCreator.areaRender = null;
+realityEditor.gui.ar.areaCreator.needsRenderUpdate = false;
+
+realityEditor.gui.ar.areaCreator.animationCallbacks = [];
+
+realityEditor.gui.ar.areaCreator.initializedPrefabs = false;
+
+realityEditor.gui.ar.areaCreator.initService = function() {
+    realityEditor.network.addPostMessageHandler('promptForArea', this.promptForAreaHandler);
+}.bind(realityEditor.gui.ar.areaCreator);
+
+realityEditor.gui.ar.areaCreator.getPrefabs = function() {
+    if (!this.initializedPrefabs) {
+        const THREE = realityEditor.gui.threejsScene.THREE;
+        const pointPillarGeometry = new THREE.CylinderGeometry(0.02, 0.02, 1, 8).translate(0, 0.5, 0).scale(1000,1000,1000);
+        const pointPillarMaterial = new THREE.MeshBasicMaterial({color: 0x66FF66, opacity: 0.6, transparent: true});
+        this.pointPillarSource = new THREE.Mesh(pointPillarGeometry, pointPillarMaterial);
+        this.wallMaterial = new THREE.MeshBasicMaterial({color: 0x66FF66, opacity: 0.5, transparent: true, side: THREE.DoubleSide, wireframe: true});
+        this.floorMaterial = new THREE.MeshBasicMaterial({color: 0x66FF66, opacity: 0.5, transparent: true, side: THREE.DoubleSide});
+        this.initializedPrefabs = true;
+    }
+    return {
+        pointPillarSource: this.pointPillarSource,
+        wallMaterial: this.wallMaterial,
+        floorMaterial: this.floorMaterial
+    }
+}.bind(realityEditor.gui.ar.areaCreator);
+
+// TODO: Fix result not being received by tool if tool loses focus
```

Comment:
Fine to leave this as a TODO for now – I'm just curious, what do you mean by the tool losing focus?

Reply:
Oops, forgot to remove that TODO. There was a bug where if you moved the camera away from the tool while defining the area, it would unload the tool (I think this is what was happening) and then the tool would never receive the callback.